secrets/field.go

```diff
@@ -0,0 +1,251 @@
+func (s Field) MarshalYAML() (interface{}, error) {
+	printRaw := currentPolicy()
+	if !printRaw {
+		return "<secret>", nil
+	}
+	return s.rawConfig, nil
+}
+
+// MarshalJSON implements the json.Marshaler interface for Field.
+func (s Field) MarshalJSON() ([]byte, error) {
+	data, err := s.MarshalYAML()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(data)
+}
+
+type mapType = map[string]any
+
+// splitProviderAndSettings separates provider-specific configuration from the generic SecretField settings.
+func splitProviderAndSettings(provReg *ProviderRegistry, baseMap mapType) (providerName string, providerData interface{}, settingsMap mapType, err error) {
+	settingsMap = make(mapType)
+
+	for k, v := range baseMap {
+		// Check if the key corresponds to a registered provider.
+		if _, err := provReg.Get(k); err == nil {
+			if providerName != "" {
+				// A provider has already been found, which is an error.
+				return "", nil, nil, fmt.Errorf("secret must contain exactly one provider type, but multiple were found: %s, %s", providerName, k)
+			}
+			providerName = k
+			providerData = v
+		} else {
+			// If it's not a provider key, treat it as a setting.
+			settingsMap[k] = v
+		}
+	}
+
+	if providerName == "" {
+		// Marshal the map back to YAML for a readable error message.
+		yamlBytes, err := yaml.Marshal(baseMap)
+		if err != nil {
+			// Fallback to the original format if marshalling fails for some reason.
+			return "", nil, nil, fmt.Errorf("no valid secret provider found in configuration: %v", baseMap)
+		}
+		return "", nil, nil, fmt.Errorf("no valid secret provider found in configuration:\n%s", string(yamlBytes))
+	}
+
+	return providerName, providerData, settingsMap, nil
+}
+
+// convertConfig takes a yaml-parsed any and unmarshals it into a typed struct.
+// It achieves this by first marshalling the input to YAML and then unmarshalling
+// it into the target struct.
+func convertConfig[T any](source any, target T) error {
+	bytes, err := yaml.Marshal(source)
+	if err != nil {
+		return fmt.Errorf("failed to re-marshal config: %w", err)
+	}
+	if err := yaml.Unmarshal(bytes, target); err != nil {
+		return fmt.Errorf("failed to unmarshal config: %w", err)
+	}
+	return nil
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface for Field.
+func (s *Field) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return unmarshal(&s.rawConfig)
+}
+
+func (s *Field) parseRawConfig(reg *ProviderRegistry, path string) (*fieldState, error) {
+	var plainSecret string
+	if err := convertConfig(s.rawConfig, &plainSecret); err == nil {
+		return &fieldState{
+			path:         path,
+			providerName: "inline",
+			config: &InlineProviderConfig{
+				secret: plainSecret,
+			},
+			value: plainSecret,
+		}, nil
+	}
+
+	var baseMap mapType
+	if err := convertConfig(s.rawConfig, &baseMap); err != nil {
+		return nil, err
+	}
+
+	providerName, providerConfigData, settingsMap, err := splitProviderAndSettings(reg, baseMap)
```

Comment:
nit: Wouldn't this be simpler if `splitProviderAndSettings` was returning `providerConfig`?